src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs

```diff
@@ -339,25 +339,36 @@ internal Task ExtractRelativeToDirectoryAsync(string destinationDirectoryPath, b
             }
 
             string? linkTargetPath = null;
-            if (EntryType is TarEntryType.SymbolicLink or TarEntryType.HardLink)
+            if (EntryType is TarEntryType.SymbolicLink)
             {
                 if (string.IsNullOrEmpty(LinkName))
                 {
                     throw new InvalidDataException(SR.TarEntryHardLinkOrSymlinkLinkNameEmpty);
                 }
 
+                // Disallow link targets that point outside the destination directory.
                 linkTargetPath = GetSanitizedFullPath(destinationDirectoryPath,
                     Path.IsPathFullyQualified(LinkName) ? LinkName : Path.Join(Path.GetDirectoryName(fileDestinationPath), LinkName));
-
                 if (linkTargetPath == null)
                 {
                     throw new IOException(SR.Format(SR.TarExtractingResultsLinkOutside, LinkName, destinationDirectoryPath));
                 }
-
-                // after TarExtractingResultsLinkOutside validation, preserve the original
-                // symlink target path (to match behavior of other utilities).
                 linkTargetPath = LinkName;
             }
+            else if (EntryType is TarEntryType.HardLink)
+            {
+                if (string.IsNullOrEmpty(LinkName))
```

Comment:
@carlossanlop LinkNames for hard links are relative paths to the archive root. We need to treat them different from symbolic link names. I've added a commit for that, and updated the existing tests accordingly.

I can move it to a separate PR if you prefer.